src/aliceVision/sfm/bundle/BundleAdjustmentCeres.cpp

```diff
@@ -446,104 +498,119 @@ void BundleAdjustmentCeres::addIntrinsicsToProblem(const sfmData::SfMData& sfmDa
-            assert(intrinsicBlock.size() >= 3);
+            if (!lockCenter)
+            {
+                // refine optical center within 10% of the image size.
+                assert(intrinsicBlock.size() >= 3);
 
-            const double opticalCenterMinPercent = -0.05;
-            const double opticalCenterMaxPercent = 0.05;
+                const double opticalCenterMinPercent = -0.05;
+                const double opticalCenterMaxPercent = 0.05;
 
-            // add bounds to the principal point
-            problem.SetParameterLowerBound(intrinsicBlockPtr, 2, opticalCenterMinPercent * intrinsicPtr->w());
-            problem.SetParameterUpperBound(intrinsicBlockPtr, 2, opticalCenterMaxPercent * intrinsicPtr->w());
-            problem.SetParameterLowerBound(intrinsicBlockPtr, 3, opticalCenterMinPercent * intrinsicPtr->h());
-            problem.SetParameterUpperBound(intrinsicBlockPtr, 3, opticalCenterMaxPercent * intrinsicPtr->h());
+                // add bounds to the principal point
+                problem.SetParameterLowerBound(intrinsicBlockPtr, 2, opticalCenterMinPercent * intrinsicPtr->w());
+                problem.SetParameterUpperBound(intrinsicBlockPtr, 2, opticalCenterMaxPercent * intrinsicPtr->w());
+                problem.SetParameterLowerBound(intrinsicBlockPtr, 3, opticalCenterMinPercent * intrinsicPtr->h());
+                problem.SetParameterUpperBound(intrinsicBlockPtr, 3, opticalCenterMaxPercent * intrinsicPtr->h());
+            }
+            
+            auto * subsetManifold = new IntrinsicsManifold(
+                                        intrinsicBlock.size(), 
+                                        focalRatio, 
+                                        lockFocal, 
+                                        lockRatio, 
+                                        lockCenter
+                                    );
+
+            problem.SetManifold(intrinsicBlockPtr, subsetManifold);
         }
 
-        // lens distortion
-        if (!refineIntrinsicsDistortion || intrinsicPtr->getDistortionInitializationMode() == camera::EInitMode::CALIBRATED)
-        {
-            lockDistortion = true;
-        }
 
         //Check if this particular distortion is locked
-        std::shared_ptr<camera::IntrinsicScaleOffsetDisto> intrinsicScaleOffsetDisto =
-              std::dynamic_pointer_cast<camera::IntrinsicScaleOffsetDisto>(intrinsicPtr);
-        if (intrinsicScaleOffsetDisto)
+        auto isod = camera::IntrinsicScaleOffsetDisto::cast(intrinsicPtr);
+        if (isod)
         {
-            if (intrinsicScaleOffsetDisto->getDistortion())
+            auto distortion = isod->getDistortion();
+            if (distortion)
             {
-                if (intrinsicScaleOffsetDisto->getDistortion()->isLocked())
+                //Create data block for intrinsics inside ceres
+                std::vector<double>& distortionBlock = _distortionsBlocks[intrinsicId];
```

Comment:
Can't `distortionBlock` be initialized straight ahead with `distortion->getParameters();` instead of `_distortionsBlocks[intrinsicId]`?

Reply:
Oh this is a tricky one here, i may have wrote some bug.
What i want (at line 597), is to copy the VALUES of distortion->getParameters() into distortionBlock.

Reply:
No it's ok. the values are copied. code ok.